lib/Support/Log.cpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Support/Log.h"
+
+namespace glow {
+
+llvm::cl::opt<bool>
+    dumpCompilationLogOpt("dump-compilation-log",
+                          llvm::cl::desc("Dump compilation log"));
+
+void LogContext::addLogContent(llvm::StringRef logContent) {
+  logContents_.push_back(logContent);
+}
+
+void LogContext::addLogScope(llvm::StringRef scopeName) {
+  logScopes_.push_back(scopeName);
+  currentFullScope_ = "";
+  for (auto s : logScopes_) {
+    currentFullScope_ += "->";
+    currentFullScope_ += s;
+  }
+}
+
+void LogContext::popBackLogScope() {
+  if (logScopes_.size() > 0) {
+    logScopes_.pop_back();
+  }
+  currentFullScope_ = "";
+  for (auto s : logScopes_) {
+    currentFullScope_ += "->";
+    currentFullScope_ += s;
+  }
+}
+void LogContext::dumpLog(llvm::StringRef funcName) {
+  if (dumpCompilationLogOpt) {
+    std::string compileLogFilename = funcName;
+    compileLogFilename.append("_compile.log");
+
+    llvm::outs() << "Writing compilation log file for Module to: "
+                 << compileLogFilename << '\n';
+    std::error_code EC;
+    llvm::raw_fd_ostream myfile(compileLogFilename, EC);
+    for (auto s : logContents_) {
```

Comment:
As far as I know, `auto` never resolves to a reference type so this implicitly makes a copy of each string in `logContents_`. Consider changing this to `const auto&` (`const` to make sure the scope name is not accidentally modified) since you don't need a local copy in this case.